cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/ConfiguratorFactory.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package io.cdap.cdap.internal.app.deploy;
+
+import io.cdap.cdap.app.deploy.Configurator;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.id.Id;
+import io.cdap.cdap.internal.app.runtime.artifact.PluginFinder;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.filesystem.Location;
+
+/**
+ * ConfiguratorFactory which returns either a RemoteConfigurator or InMemoryConfigurator
+ */
+public class ConfiguratorFactory {
```

Comment:
I'm not sure if this will work since RemoteConfigurator and InMemoryConfigurator require a lot of constructor parameters that cant be injected, they are only available at runtime in the `LocalArtifactLoaderStage`. We initially added the factory (instead of doing the binding directly) because we needed to be able to pass in the parameters at runtime. 

Is there a way to do this with just a provider?

Reply:
I spoke with Terence offline, he recommended using the `@Assisted` annotation for Guice. 

I looked into it but it seems like its a non-trivial change since our binding depends on a value from within the CConfiguration. All existing usages of `@Assisted` and `FactoryModuleBuilder()` rely on a fixed mapping while ours is dynamic. I'll leave this the way it is for now and we can fix this in a follow up PR